### PR TITLE
Rework handling of oversize text messages.

### PR DESCRIPTION
**Pull request by charlesmchen**
PTAL @michaelkirk

### CODE REVIEW

**Comment by charlesmchen** [comment]
screen shot 2017-10-25 at 2 57 15 pm
screen shot 2017-10-25 at 2 57 03 pm

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m [hunk 1]
```diff
@@ -130,7 +131,6 @@ - (void)commontInit
     self.textView = [UITextView new];
     // Honor dynamic type in the message bodies.
     self.textView.font = [self textMessageFont];
-    self.textView.font = [UIFont ows_regularFontWithSize:16.f];
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: I don't know if this was rebase error, but this was probably breaking dynamic type handling.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m [hunk 2]
```diff
@@ -183,17 +183,27 @@ - (UIFont *)textMessageFont
...
+    return (CGFloat)ceil([self tapForMoreFont].lineHeight * 1.25);
+}
+
 - (OWSMessageCellType)cellType
 {
     return self.viewItem.messageCellType;
 }
 
-- (nullable NSString *)textMessage
+- (nullable DisplayableText *)displayableText
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageCell.m]: I've introduced `DisplayableText` which has a text messages:

* Full, stripped and de-zalgo-ized text.
* Displayable (e.g. possibly truncated if very long) stripped de-zalgo-ized text.
* A flag indicating whether or not it was truncated (tap for more...).

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 3]
```diff
@@ -2024,18 +2024,17 @@ - (void)didTapAudioViewItem:(ConversationViewItem *)viewItem attachmentStream:(T
...
+- (void)didTapTruncatedTextMessage:(ConversationViewItem *)conversationItem
 {
     OWSAssert([NSThread isMainThread]);
-    OWSAssert(displayableText);
-    OWSAssert(attachmentStream);
+    OWSAssert(conversationItem);
 
-    // Tapping on incoming and outgoing "oversize text messages" should show the
-    // "oversize text message" view.
-    OversizeTextMessageViewController *messageVC =
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: Got rid of `OversizeTextMessageViewController` in favor of `MessageMetadataViewController`

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 4]
```diff
@@ -2024,18 +2024,17 @@ - (void)didTapAudioViewItem:(ConversationViewItem *)viewItem attachmentStream:(T
...
-        [[OversizeTextMessageViewController alloc] initWithDisplayableText:displayableText
-                                                          attachmentStream:attachmentStream];
-    [self.navigationController pushViewController:messageVC animated:YES];
+    TSMessage *message = (TSMessage *)conversationItem.interaction;
+    MessageMetadataViewController *view =
+        [[MessageMetadataViewController alloc] initWithViewItem:conversationItem
+                                                        message:message
+                                                           mode:MessageMetadataViewModeFocusOnMessage];
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: Notice the "focus on" argument.

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.m [hunk 5]
```diff
@@ -266,49 +274,71 @@ - (NSCache *)displayableTextCache
...
+                                           textBlock:(NSString * (^_Nonnull)())textBlock
+{
+    OWSAssert(interactionId.length > 0);
+
+    DisplayableText *_Nullable displayableText = [[self displayableTextCache] objectForKey:interactionId];
     if (!displayableText) {
+        NSString *text = textBlock();
+
         // Only show up to 2kb of text.
         const NSUInteger kMaxTextDisplayLength = 2 * 1024;
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: I've unified the "strip, de-zalgo and possibly truncate" logic here.

Comment (charlesmchen) [reply Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: This also will cut down on a bunch of redundant (and possibly expensive) work by leaning on the cache more.

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.m [hunk 6]
```diff
@@ -495,7 +529,8 @@ - (void)copyAction
     switch (self.messageCellType) {
         case OWSMessageCellType_TextMessage:
         case OWSMessageCellType_OversizeTextMessage:
-            [UIPasteboard.generalPasteboard setString:self.textMessage];
+            OWSAssert(self.displayableText);
+            [UIPasteboard.generalPasteboard setString:self.displayableText.fullText];
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: Fixed bugs around us sharing or copying the truncated text.

File: Signal/src/ViewControllers/CropScaleImageViewController.swift [hunk 7]
```diff
@@ -101,15 +101,9 @@ class CropScaleImageViewController: OWSViewController {
...
+    @available(*, unavailable, message:"use other constructor instead.")
     required init?(coder aDecoder: NSCoder) {
-        self.srcImage = UIImage(named:"fail")!
-        self.successCompletion = { _ in
-        }
-        super.init(coder: aDecoder)
-        owsFail("\(self.TAG) invalid constructor")
-
-        configureCropAndScale()
+        fatalError("\(#function) is unimplemented.")
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/CropScaleImageViewController.swift]: Love `fatalError`

File: Signal/src/ViewControllers/MessageMetadataViewController.swift [hunk 8]
```diff
@@ -267,76 +276,79 @@ class MessageMetadataViewController: OWSViewController {
         }
     }
 
+    private func displayableTextIfText() -> String? {
+        let messageCellType = viewItem.messageCellType()
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/MessageMetadataViewController.swift]: We weren't doing the "strip, de-zalgo" thing here.  By using `ConversationViewItem`, we can lean on its cache and DRY up this logic.

File: Signal/src/ViewControllers/MessageMetadataViewController.swift [hunk 9]
```diff
@@ -267,76 +276,79 @@ class MessageMetadataViewController: OWSViewController {
...
+
+            let isIncoming = self.message as? TSIncomingMessage != nil
+
+            // UITextView can't render extremely long text due to constraints
+            // on the size of its backing buffer, especially when we're 
+            // embedding it "full-size' within a UIScrollView as we do in this view.
+            //
+            // TODO: We could use CoreText instead, or we could dynamically
+            //       manipulate the size/position of our UITextView to
+            //       reflect scroll state.
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/MessageMetadataViewController.swift]: This is a serious issue.  I'm going to make a card.  I'm not sure of its priority.

File: Signal/src/ViewControllers/MessageMetadataViewController.swift [hunk 10]
```diff
@@ -4,6 +4,12 @@
 
 import Foundation
 
+@objc
+enum MessageMetadataViewMode: UInt {
+    case focusOnMessage
+    case focusOnMetadata
+}
+
 class MessageMetadataViewController: OWSViewController {
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/MessageMetadataViewController.swift]: I propose that we rename `MessageMetadataViewController` as `SingleMessageViewController`.

File: Signal/src/ViewControllers/OversizeTextMessageViewController.swift [hunk 11]
```diff
@@ -1,75 +0,0 @@
-//
-//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
-//
-
-import Foundation
-import WebRTC
-import PromiseKit
-
-class OversizeTextMessageViewController: OWSViewController {
```
Comment (charlesmchen) [inline Signal/src/ViewControllers/OversizeTextMessageViewController.swift]: Goodbye!

File: Signal/src/ViewControllers/MessageMetadataViewController.swift (hunk 10 above)
Comment (michaelkirk) [reply Signal/src/ViewControllers/MessageMetadataViewController.swift]: :+1: Or maybe `MessageDetailViewController` or `MessageInfoViewController`

Comment (charlesmchen) [reply Signal/src/ViewControllers/MessageMetadataViewController.swift]: MessageDetailViewController

**Comment by charlesmchen** [comment]
> If it's simple, we could keep our class count down by combining the DisplayableText and DisplayableTextFilter classes into a single class, but no biggie.

Done.

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 12]
```diff
@@ -2024,18 +2024,17 @@ - (void)didTapAudioViewItem:(ConversationViewItem *)viewItem attachmentStream:(T
...
-    OWSAssert(attachmentStream);
+    OWSAssert(conversationItem);
 
-    // Tapping on incoming and outgoing "oversize text messages" should show the
-    // "oversize text message" view.
-    OversizeTextMessageViewController *messageVC =
-        [[OversizeTextMessageViewController alloc] initWithDisplayableText:displayableText
-                                                          attachmentStream:attachmentStream];
-    [self.navigationController pushViewController:messageVC animated:YES];
+    TSMessage *message = (TSMessage *)conversationItem.interaction;
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: Done.

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.m [hunk 13]
```diff
@@ -266,49 +274,71 @@ - (NSCache *)displayableTextCache
...
+    DisplayableText *_Nullable displayableText = [[self displayableTextCache] objectForKey:interactionId];
     if (!displayableText) {
+        NSString *text = textBlock();
+
         // Only show up to 2kb of text.
         const NSUInteger kMaxTextDisplayLength = 2 * 1024;
         text = [text ows_stripped];
-        displayableText = [[DisplayableTextFilter new] displayableText:text];
-        if (displayableText.length > kMaxTextDisplayLength) {
+        NSString *fullText = [[[DisplayableTextFilter new] displayableText:text] ows_stripped];
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply Signal/src/ViewControllers/ConversationView/ConversationViewItem.m]: Done.

**Pull request merged**